src/hotspot/share/opto/reachability.cpp

```diff
@@ -0,0 +1,555 @@
+  // Also, newly introduced RF can make some existing RFs redundant.
+  find_redundant_rfs(redundant_rfs);
+
+  // Eliminate redundant RFs.
+  bool progress = (redundant_rfs.size() > 0);
+  while (redundant_rfs.size() > 0) {
+    remove_rf(redundant_rfs.pop()->as_ReachabilityFence());
+  }
+
+  assert(redundant_rfs.size() == 0, "");
+  assert(!has_redundant_rfs(redundant_rfs, true /*rf_only*/), "");
+
+  return progress;
+}
+
+//======================================================================
+//---------------------------- Phase 2 ---------------------------------
+
+// Linearly traverse CFG upwards starting at n until first merge point.
+// All encountered safepoints are recorded in safepoints list.
+static void linear_traversal(Node* n, Node_Stack& worklist, VectorSet& visited, Node_List& safepoints) {
+  for (Node* ctrl = n; ctrl != nullptr; ctrl = ctrl->in(0)) {
+    assert(ctrl->is_CFG(), "");
+    if (visited.test_set(ctrl->_idx)) {
+      return;
+    } else {
+      if (ctrl->is_Region()) {
+        worklist.push(ctrl, 1);
+        return; // stop at merge points
+      } else if (ctrl->is_SafePoint() && is_interfering_sfpt_candidate(ctrl->as_SafePoint())) {
+        safepoints.push(ctrl);
+      }
+    }
+  }
+}
+
+// Enumerate all safepoints which are reachable from the RF to its referent through CFG.
+// Start at RF node and traverse CFG upwards until referent's control node is reached.
+static void enumerate_interfering_sfpts(ReachabilityFenceNode* rf, PhaseIdealLoop* phase, Node_List& safepoints) {
+  Node* referent = rf->referent();
+  Node* referent_ctrl = phase->get_ctrl(referent);
+  assert(phase->is_dominator(referent_ctrl, rf), "sanity");
+
+  VectorSet visited;
+  visited.set(referent_ctrl->_idx); // end point
+
+  Node_Stack stack(0);
+  linear_traversal(rf, stack, visited, safepoints); // start point
+  while (stack.is_nonempty()) {
+    Node* cur = stack.node();
+    uint  idx = stack.index();
+
+    assert(cur != nullptr, "");
+    assert(cur->is_Region(), "%s", NodeClassNames[cur->Opcode()]);
+    assert(phase->is_dominator(referent_ctrl, cur), "");
+    assert(idx > 0 && idx <= cur->req(), "%d %d", idx, cur->req());
+
+    if (idx < cur->req()) {
+      stack.set_index(idx + 1);
+      linear_traversal(cur->in(idx), stack, visited, safepoints);
+    } else {
+      stack.pop();
+    }
+  }
+}
+
+// Start offset for reachability info on a safepoint node.
+static uint rf_base_offset(SafePointNode* sfpt) {
+  return sfpt->jvms()->oopoff();
+}
+
+// Phase 2: migrate reachability info to safepoints.
+// All RFs are replaced with edges from corresponding referents to interfering safepoints.
+// Interfering safepoints are safepoint nodes which are reachable from the RF to its referent through CFG.
+bool PhaseIdealLoop::eliminate_reachability_fences() {
+  Compile::TracePhase tp(_t_reachability_eliminate);
+
+  assert(OptimizeReachabilityFences, "required");
+  assert(C->post_loop_opts_phase(), "required");
```

Comment:
You could give the reason in the assert message ;)

Reply:
Those are essentially preconditions. I'm not a fan of duplicating information over and over again. Assert messages won't add much beyond what the condition already says unless the message explains why precondition is there. But that would explode message size.

I can use `precond` here if you are against generic messages in asserts.